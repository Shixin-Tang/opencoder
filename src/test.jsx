import React, { useState, useEffect } from "react"
import { render, Box, Text } from "ink2"
import Static from "./components/static.js"
import { render as renderInk } from "../test/testutil"

const Example = () => {
  const [tests, setTests] = useState([])

  useEffect(() => {
    let completedTests = 0
    let timer

    const run = () => {
      // Fake 10 completed tests
      if (completedTests++ < 10) {
        setTests((previousTests) => [
          ...previousTests,
          {
            id: previousTests.length,
            title: `Test #${previousTests.length + 1}`,
          },
        ])

console.log(renderInk(<Text color="green">✔Hello</Text>).lastFrame())

        timer = setTimeout(run, 100)
      }
    }

    run()

    return () => {
      clearTimeout(timer)
    }
  }, [])

  return (
    <Box flexDirection="column" gap={1} position="relative">
      {/* This part will be rendered once to the terminal */}
      <Static items={tests}>
        {(test) => (
          <Box key={test.id}>
            <Text color="green">✔ {test.title}</Text>
          </Box>
        )}
      </Static>

      {/* This part keeps updating as state changes */}
      <Box marginTop={1}>
        <Text dimColor>Completed tests: {tests.length}</Text>
      </Box>
    </Box>
  )
}

render(<Example />)
