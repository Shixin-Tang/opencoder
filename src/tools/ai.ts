import type { ReactNode } from "react"
import { type JSONValue, type Tool, type ToolInvocation } from "ai"

// TODO: annotate message with tool output params, eg file patch, useful for persistent chat messages.

type Streamable = ReactNode | Promise<ReactNode> | JSONValue

export type Renderer<T extends Array<any>> = (...args: T) => AsyncGenerator<Streamable, void, void>

export type CoderTool = Omit<Tool, "execute"> & {
  generate?: Renderer<any>
  renderTitle?: (part: ToolInvocation) => ReactNode
  render?: (part: ToolInvocation) => ReactNode
  execute?: Tool["execute"]
}

export function defineTool(params: CoderTool): CoderTool {
  return params
}
