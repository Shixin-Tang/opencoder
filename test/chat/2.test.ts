import { tool } from "ai"
import { convertArrayToReadableStream, MockLanguageModelV1 } from "ai/test"

import { z } from "zod"
import { buildComponentTree } from "../utils/debugger.js"
import { queryComponentTree } from "../utils/query.js"
import { waitNextRender } from "../utils/render.js"
import { setupTestEnvironment } from "./util.js"

test("chat interaction with a successful tool call (e.g., file edit)", async () => {
  // 1. Setup: Mock model for tool request, mock MCP tool, then mock final response
  const mockEditFileTool = vi
    .fn()
    .mockResolvedValue({ success: true, message: "File edited successfully." })
  const tools = {
    edit_file: tool({
      execute: mockEditFileTool,
      description: "Edits a file",
      parameters: z.object({
        path: z.string(),
        content: z.string(),
      }),
    }),
  }

  const mockModel = new MockLanguageModelV1({})
  const doStreamMock = vi.fn()
  mockModel.doStream = doStreamMock

  // First call: Request the tool
  doStreamMock.mockImplementationOnce(async () => ({
    stream: convertArrayToReadableStream([
      {
        type: "tool-call",
        toolCallId: "tool_123",
        toolName: "edit_file",
        args: JSON.stringify({ path: "test.txt", content: "new content" }),
      },
      {
        type: "finish",
        finishReason: "tool-calls",
        usage: { promptTokens: 10, completionTokens: 5 },
      },
    ]),
    rawCall: { rawPrompt: null, rawSettings: {} },
    type: "finish",
    finishReason: "tool-calls",
    usage: { promptTokens: 10, completionTokens: 5 },
  }))

  // Second call: Respond after successful tool execution
  doStreamMock.mockImplementationOnce(async ({ prompt }) => {
    // Basic check: Verify the tool result is in the prompt
    const promptMessages = prompt
    const lastMessage = promptMessages[promptMessages.length - 1]
    expect(lastMessage.role).toBe("tool")
    expect(lastMessage.content).toEqual(
      JSON.stringify({ success: true, message: "File edited successfully." }),
    )
    expect(lastMessage.toolCallId).toBe("tool_123")

    return {
      stream: convertArrayToReadableStream([
        { type: "text-delta", textDelta: "OK, " },
        { type: "text-delta", textDelta: "I've edited the file." },
        { type: "finish", finishReason: "stop", usage: { promptTokens: 25, completionTokens: 8 } },
      ]),
      rawCall: { rawPrompt: JSON.stringify(prompt), rawSettings: {} },
      type: "finish",
      finishReason: "stop",
      usage: { promptTokens: 25, completionTokens: 8 },
    }
  })

  const { instance, stdin, stdout, fiber, queryClient, store, config } = await setupTestEnvironment(
    {
      model: mockModel,
      customTools: tools,
    },
  )

  expect(stdout.get()).toMatchSnapshot("mcp tool - initial")
  assert(stdin)

  // 2. Simulate user input
  stdin.emit("input", "Please edit test.txt")
  await waitNextRender()
  stdin.emit("input", "\r")
  await waitNextRender()

  // 3. Wait for MCP UI & Verify
  // Use waitFor as the UI update might take a moment after the stream finishes
  await vi.waitFor(
    () => {
      expect(stdout.get()).toContain("Custom tool: edit_file")
      expect(stdout.get()).toContain("File edited successfully")
    },
    { timeout: 500 },
  )
  expect(stdout.get()).toMatchSnapshot("mcp tool - confirmation prompt")

  // 4. Simulate user approval (assuming 'y' or Enter confirms)
  stdin.emit("input", "\r")
  await waitNextRender()
  await waitNextRender()
  await waitNextRender()

  // 5. Wait for final response rendering
  await vi.waitFor(
    () => {
      const tree = buildComponentTree(fiber!.current.child)
      if (queryComponentTree(tree, "Spinner")) {
        throw new Error("Spinner/Loading indicator still present")
      }
    },
    { timeout: 1000 },
  )

  // 6. Verify tool was called
  expect(mockEditFileTool).toHaveBeenCalledTimes(1)
  expect(mockEditFileTool.mock.calls[0][0]).toEqual({ path: "test.txt", content: "new content" })

  // 7. Verify model received tool result (checked inside mock implementation)
  expect(doStreamMock).toHaveBeenCalledTimes(2)

  // 8. Final state check
  expect(stdout.get()).toMatchSnapshot("mcp tool - final state")

  instance.unmount()
})
