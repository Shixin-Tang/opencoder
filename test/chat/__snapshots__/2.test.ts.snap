// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`chat interaction with a successful tool call (e.g., file edit) > mcp tool - confirmation prompt 1`] = `
"> Please edit test.txt
▼  Custom tool: edit_file
 ╭─────────────────────────────────────────────────────────────────────────────────────────────────╮
 │{ success: [33mtrue[39m, message: [32m'File edited successfully.'[39m }                                          │
 ╰─────────────────────────────────────────────────────────────────────────────────────────────────╯

╭──────────────────────────────────────────────────────────────────────────────────────────────────╮
│ An error occurred while processing your request:                                                 │
│                                                                                                  │
│ expected [ { type: 'tool-result', …(6) } ] to deeply equal '{"success":true,"message":"File      │
│ edite…'                                                                                          │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯


╭──────────────────────────────────────────────────────────────────────────────────────────────────╮
│ >                                                                                                │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
  / for commands                                                    alt + ⏎ for newline ·15 tokens"
`;

exports[`chat interaction with a successful tool call (e.g., file edit) > mcp tool - final state 1`] = `
"> Please edit test.txt
▼  Custom tool: edit_file
 ╭─────────────────────────────────────────────────────────────────────────────────────────────────╮
 │{ success: [33mtrue[39m, message: [32m'File edited successfully.'[39m }                                          │
 ╰─────────────────────────────────────────────────────────────────────────────────────────────────╯

╭──────────────────────────────────────────────────────────────────────────────────────────────────╮
│ An error occurred while processing your request:                                                 │
│                                                                                                  │
│ expected [ { type: 'tool-result', …(6) } ] to deeply equal '{"success":true,"message":"File      │
│ edite…'                                                                                          │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯


╭──────────────────────────────────────────────────────────────────────────────────────────────────╮
│ >                                                                                                │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
  / for commands                                                    alt + ⏎ for newline ·15 tokens"
`;

exports[`chat interaction with a successful tool call (e.g., file edit) > mcp tool - initial 1`] = `
"
╭──────────────────────────────────────────────────────────────────────────────────────────────────╮
│ >                                                                                                │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
  / for commands                                                     alt + ⏎ for newline ·0 tokens"
`;
